Application/EdFi.Ods.Extensions.TPDM/Artifacts/MsSql/Data/Security/1010-TPDM-ResourceClaims.sql

```diff
@@ -1,465 +1,6004 @@
+    DELETE FROM dbo.ResourceClaimAuthorizationMetadatas
+    WHERE ResourceClaim_ResourceClaimId = @claimId
+    
+    -- Default Create authorization
+    SET @authorizationStrategyId = NULL
+
+    SELECT @authorizationStrategyId = a.AuthorizationStrategyId
+    FROM    dbo.AuthorizationStrategies a
+    WHERE   a.DisplayName = 'Relationships with Education Organizations only'
+
+    IF @authorizationStrategyId IS NULL
+    BEGIN
+        SET @msg = 'AuthorizationStrategy does not exist: ''Relationships with Education Organizations only''';
+        THROW 50000, @msg, 1
+    END
+
+    INSERT INTO dbo.ResourceClaimAuthorizationMetadatas(ResourceClaim_ResourceClaimId, Action_ActionId, AuthorizationStrategy_AuthorizationStrategyId)
+    VALUES (@claimId, @CreateActionId, @authorizationStrategyId)
+
+    -- Default Read authorization
+    SET @authorizationStrategyId = NULL
+
+    SELECT @authorizationStrategyId = a.AuthorizationStrategyId
+    FROM    dbo.AuthorizationStrategies a
+    WHERE   a.DisplayName = 'Relationships with Education Organizations only'
+
+    IF @authorizationStrategyId IS NULL
+    BEGIN
+        SET @msg = 'AuthorizationStrategy does not exist: ''Relationships with Education Organizations only''';
+        THROW 50000, @msg, 1
+    END
+
+    INSERT INTO dbo.ResourceClaimAuthorizationMetadatas(ResourceClaim_ResourceClaimId, Action_ActionId, AuthorizationStrategy_AuthorizationStrategyId)
+    VALUES (@claimId, @ReadActionId, @authorizationStrategyId)
+
+    -- Default Update authorization
+    SET @authorizationStrategyId = NULL
+
+    SELECT @authorizationStrategyId = a.AuthorizationStrategyId
+    FROM    dbo.AuthorizationStrategies a
+    WHERE   a.DisplayName = 'Relationships with Education Organizations only'
+
+    IF @authorizationStrategyId IS NULL
+    BEGIN
+        SET @msg = 'AuthorizationStrategy does not exist: ''Relationships with Education Organizations only''';
+        THROW 50000, @msg, 1
+    END
+
+    INSERT INTO dbo.ResourceClaimAuthorizationMetadatas(ResourceClaim_ResourceClaimId, Action_ActionId, AuthorizationStrategy_AuthorizationStrategyId)
+    VALUES (@claimId, @UpdateActionId, @authorizationStrategyId)
+
+    -- Default Delete authorization
+    SET @authorizationStrategyId = NULL
+
+    SELECT @authorizationStrategyId = a.AuthorizationStrategyId
+    FROM    dbo.AuthorizationStrategies a
+    WHERE   a.DisplayName = 'Relationships with Education Organizations only'
+
+    IF @authorizationStrategyId IS NULL
+    BEGIN
+        SET @msg = 'AuthorizationStrategy does not exist: ''Relationships with Education Organizations only''';
+        THROW 50000, @msg, 1
+    END
+
+    INSERT INTO dbo.ResourceClaimAuthorizationMetadatas(ResourceClaim_ResourceClaimId, Action_ActionId, AuthorizationStrategy_AuthorizationStrategyId)
+    VALUES (@claimId, @DeleteActionId, @authorizationStrategyId)
+
+    -- Processing claim sets for http://ed-fi.org/ods/identity/claims/tpdm/teacherPreparationProviderProgram
```

Comment:
I thought this was strange too, but here's the back story.

I was seeing an error in the TPDM template log where it indicated that the API client was missing a claim, but the message didn't provide an indication of what claim set was actually in use. So the changes in the Ed-Fi-ODS repository for this ticket enhance the error message so that it now includes the API client's assigned Claim Set.

As it turns out, this resource is being loaded under that "Bootstrap" claim set, as the following improved error message indicates:
```
EdFi.Ods.Api.Services.Controllers.TPDM.TeacherPreparationProviderPrograms.TeacherPreparationProviderProgramsController: 2020-07-18 19:42:50,936 [13] ERROR EdFi.Ods.Api.Services.Controllers.TPDM.TeacherPreparationProviderPrograms.Teacher
EdFi.Ods.Common.Security.EdFiSecurityException: Access to the resource could not be authorized. Are you missing a claim? This resource can be authorized by the following claims:
    http://ed-fi.org/ods/identity/claims/tpdm/teacherPreparationProviderProgram
    http://ed-fi.org/ods/identity/claims/domains/tpdm/noFurtherAuthorizationRequiredData
    http://ed-fi.org/ods/identity/claims/domains/tpdm
The API client has been assigned the 'Bootstrap Descriptors and EdOrgs' claim set with the following resource claims:
    http://ed-fi.org/ods/identity/claims/domains/systemDescriptors
    http://ed-fi.org/ods/identity/claims/domains/managedDescriptors
    http://ed-fi.org/ods/identity/claims/domains/educationOrganizations
    http://ed-fi.org/ods/identity/claims/accountabilityRating
    http://ed-fi.org/ods/identity/claims/classPeriod
    http://ed-fi.org/ods/identity/claims/communityOrganization
    http://ed-fi.org/ods/identity/claims/communityProvider
    http://ed-fi.org/ods/identity/claims/communityProviderLicense
    http://ed-fi.org/ods/identity/claims/course
    http://ed-fi.org/ods/identity/claims/educationOrganizationNetwork
    http://ed-fi.org/ods/identity/claims/educationOrganizationNetworkAssociation
    http://ed-fi.org/ods/identity/claims/educationOrganizationPeerAssociation
    http://ed-fi.org/ods/identity/claims/educationServiceCenter
    http://ed-fi.org/ods/identity/claims/feederSchoolAssociation
    http://ed-fi.org/ods/identity/claims/learningObjective
    http://ed-fi.org/ods/identity/claims/learningStandard
    http://ed-fi.org/ods/identity/claims/learningStandardEquivalenceAssociation
    http://ed-fi.org/ods/identity/claims/localEducationAgency
    http://ed-fi.org/ods/identity/claims/location
    http://ed-fi.org/ods/identity/claims/postSecondaryInstitution
    http://ed-fi.org/ods/identity/claims/program
    http://ed-fi.org/ods/identity/claims/school
    http://ed-fi.org/ods/identity/claims/stateEducationAgency
   at EdFi.Ods.Security.Authorization.EdFiAuthorizationProvider.GetAuthorizationDetails(EdFiAuthorizationContext authorizationContext)
   at EdFi.Ods.Security.Authorization.EdFiAuthorizationProvider.<AuthorizeSingleItemAsync>d__7.MoveNext() in D:\buildagent\work\7540bc60aa3a490f\Ed-Fi-ODS\Application\EdFi.Ods.Security\Authorization\EdFiAuthorizationProvider.cs:line 100
```
So, that's why the Create permission has been added for the Bootstrap claim set -- because the TPDM load process needed it to be.

If you look at the security metadata that was in place before my work, you'll find that _teacherPreparationProviderProgram_ resource claim was actually located under _educationOrganizations_ (along with a bunch of other non-EdOrg resources, see [ODS-4380](https://tracker.ed-fi.org/browse/ODS-4380)) and with explicit default authorization metadata using "No Further Auth" for all operations (which I don't think is actually used because there is an override for the "Bootstrap" claim set on the Create operation on educationOrganizations that takes precedence). In any case, the existing metadata has been cleaned up, but the claim set in use by the TPDM load process necessitated this particular setting.

Reply:
This is a bug with the TPDM load process then, it shouldn't be loading these as part of bootstrap.  We can spin up another ticket to look into that.  Bootstrap should be as limited as possible.